src/renderers/shared/stack/reconciler/ReactCompositeComponent.js

```diff
@@ -408,47 +395,35 @@ var ReactCompositeComponentMixin = {
     updateQueue
   ) {
     var Component = this._currentElement.type;
-    var instanceOrElement;
+
+    var debugID = 0;
+    if (__DEV__) {
+      debugID = this._debugID;
+    }
+
     if (doConstruct) {
-      if (__DEV__) {
-        if (this._debugID !== 0) {
-          ReactInstrumentation.debugTool.onBeginLifeCycleTimer(
-            this._debugID,
-            'ctor'
-          );
-        }
-      }
-      instanceOrElement = new Component(publicProps, publicContext, updateQueue);
-      if (__DEV__) {
-        if (this._debugID !== 0) {
-          ReactInstrumentation.debugTool.onEndLifeCycleTimer(
-            this._debugID,
-            'ctor'
-          );
-        }
+      if (__DEV__ && debugID !== 0) {
```

Comment:
Valid question 😄 
It's used for two weird internal kinds of components:

* "Top level wrappers" which aren't real components, but things we use to handle top-level `render()`s as updates when the type matches. This is an ugly implementation detail that ideally should be removed eventually.

* "Empty components" which also aren't real components, but that we mount as a child when you return `null` from render. Again, these exist only as implementation detail, and aren't exposed to user in any way.

Neither of these two "pseudo" components should be exposed to ReactPerf, DevTools, or hooks, so for now I gave them `debugID = 0`. They wouldn't exist in Fiber, or even in current reconciler if we refactor some things.